lib/chewy/strategy/lazy_sidekiq.rb

```diff
@@ -0,0 +1,54 @@
+module Chewy
+  class Strategy
+    # The strategy works the same way as sidekiq, but performs
+    # async evaluation of all index callbacks on model create and update
+    # driven by sidekiq
+    #
+    #   Chewy.strategy(:lazy_sidekiq) do
+    #     User.all.map(&:save) # Does nothing here
+    #     Post.all.map(&:save) # And here
+    #     # It imports all the changed users and posts right here
+    #   end
+    #
+    class LazySidekiq < Sidekiq
+      class LazyWorker
+        include ::Sidekiq::Worker
+
+        def perform(type, id)
+          type.constantize.find_by_id(id)&.run_chewy_callbacks
+        end
+      end
+
+      def initialize
+        super
+
+        @stash = []
+      end
+
+      def update(type, objects, _options = {})
+        ids = type.root.id ? Array.wrap(objects) : type.adapter.identify(objects)
+        return if ids.empty?
+
+        ::Sidekiq::Client.push(
+          'queue' => sidekiq_queue,
+          'class' => Chewy::Strategy::Sidekiq::Worker,
+          'args'  => [type.name, ids]
+        )
+      end
+
+      def leave
+        @stash.each do |model_name, id|
+          ::Sidekiq::Client.push(
+            'queue' => sidekiq_queue,
+            'class' => Chewy::Strategy::LazySidekiq::LazyWorker,
+            'args'  => [model_name, id]
+          )
+        end
+      end
```

Comment:
What happens if we update multiple objects? Do we run callback for them in separate jobs? If it's the case let's consider the situation when the collection is updated when callback are run. We may end up with some of the objects indexed with the intended state (with the change that scheduled the indexing) and some of them - with the new state. If we index with every update it's not a problem, but if we index only some updates, we'll end up with inconsistent state in  elastic.

Example:
```
Chewy.strategy(:lazy_sidekiq) do
  User.find(1).update!(score: 100, name: 'John')
  User.find(2).update!(score: 200, name: 'Luke')  
  # We're scheduling lazy sidekiq jobs here
end

# When the jobs are running, we're updating  records without sidekiq
 User.find(1).update!(score: 500)  
 User.find(2).update!(score: 400)  
 User.find(3).update!(score: 200)  
```
It may happen that we index
```
{`name`: `John`, 'score': 100}
{`name`: `Luke`, 'score': 400}
```
I'm not sure if we should care about this scenario, but I wanted to point think it's good to at  least think about it before merging.

Reply:
@mrzasa Currently in this strategy we schedule different jobs for every updated record within strategy block. And then when the job is performed it schedules a job for every callback updating index.

I think I got your concern. We could schedule only one job in the end of the strategy block with an array of type/id tuples of records that should be updated, .e.g. `[['User', 1], ['User', 2]]`. Then when the job run we can run all the chewy callbacks on records we have and schedule a new job for every index, just like we do in `Sidekiq` strategy. Would it solve that possible issue?

Reply:
Yes, we're thinking about the same issue and yes- I think that scheduling only one job should help.

Reply:
@mrzasa should be fixed now